python/lsst/meas/astrom/matchPessimisticB.py

```diff
@@ -0,0 +1,594 @@
+        are suitable.
+
+        @param[in] refCat  catalog of position reference stars that overlap an
+            exposure
+        @param[in] sourceCat  catalog of sources found on the exposure
+        @param[in] wcs  estimated WCS of exposure
+        @param[in] refFluxField  field of refCat to use for flux
+        @param[in] numUsableSources  number of usable sources (sources with
+            known centroid that are not near the edge, but may be saturated)
+        @param[in] minMatchedPairs  minimum number of matches
+        @param[in] match_tolerance a MatchTolerance object containing
+            variables specifying matcher tolerances and state from possible
+            previous runs.
+        @param[in] sourceInfo  SourceInfo for the sourceCat
+        @param[in] verbose  true to print diagnostic information to std::cout
+
+        @return a list of matches, an instance of
+            lsst.afw.table.ReferenceMatch, a MatchTolerance object
+        """
+
+        # Set configurable defaults when we encounter None type or set
+        # state based on pervious run of matchandfit.
+        if match_tolerance.maxShift is None:
+            maxShiftArcseconds = (self.config.maxOffsetPix *
+                                  wcs.pixelScale().asArcseconds())
+        else:
+            maxShiftArcseconds = np.max(
+                (match_tolerance.maxShift.asArcseconds(),
+                 wcs.pixelScale().asArcseconds()))
+
+        # Load the source and reference catalog as spherical points
+        # in numpy array. We do this rather than relying on internal
+        # lsst C objects for simplicity and because we require
+        # objects contiguous in memory.
+        ref_array = np.empty((len(refCat), 4))
+        for ref_idx, refObj in enumerate(refCat):
+            theta = np.pi/2 - refObj.getDec().asRadians()
+            phi = refObj.getRa().asRadians()
+            ref_array[ref_idx, 0] = np.sin(theta)*np.cos(phi)
+            ref_array[ref_idx, 1] = np.sin(theta)*np.sin(phi)
+            ref_array[ref_idx, 2] = np.cos(theta)
+            ref_array[ref_idx, 3] = -2.5*np.log10(
+                np.where(refObj[refFluxField] > 0,
+                         refObj[refFluxField], 10**-32))
+
+        src_array = np.empty((len(sourceCat), 4))
+        for src_idx, srcObj in enumerate(sourceCat):
+            coord = wcs.pixelToSky(srcObj.getCentroid())
+            tmp_ra = coord.getLongitude().asRadians()
+            tmp_dec = coord.getLatitude().asRadians()
+            theta = np.pi/2 - tmp_dec
+            phi = tmp_ra
+            src_array[src_idx, 0] = np.sin(theta)*np.cos(phi)
+            src_array[src_idx, 1] = np.sin(theta)*np.sin(phi)
+            src_array[src_idx, 2] = np.cos(theta)
+            src_array[src_idx, 3] = -2.5*np.log10(
+                np.where(srcObj.getPsfFlux() > 0,
+                         srcObj.getPsfFlux(), 10**-32))
```

Comment:
Abstracted the flux and xyz conversion into a function. Accessors to flux and RA, DEC are not consistent between ReferenceCatalog and SourceCatalog so those calls remain.